<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{pQTLtools}
-->

An Introduction to **pQTLtools** Package
========================================

```{r setup, include=FALSE}
set.seed(0)
knitr::opts_chunk$set(
  out.extra = 'style="display:block; margin: auto"',
  fig.align = "center",
  fig.path = "figures/",
  dev = "png")
```

First, we have cis/trans classification,
```{r}
require(gap)
INF <- Sys.getenv("INF")
f <- file.path(INF,"work","INF1.merge")
clumped <- read.delim(f,as.is=TRUE)
hits <- merge(clumped[c("CHR","POS","MarkerName","prot","log10p")],
              inf1[c("prot","uniprot")],by="prot")
names(hits) <- c("prot","Chr","bp","SNP","log10p","uniprot")
cistrans <- cis.vs.trans.classification(hits,inf1,"uniprot")
cis.vs.trans <- with(cistrans,data)
with(cistrans,table)
with(cistrans,total)
circos.cis.vs.trans.plot(hits=f,inf1,"uniprot")
```
This is used to generate 2d-Manhattan plot
```{r}
INF <- Sys.getenv("INF")
d <- read.csv(file.path(INF,"work","INF1.merge.cis.vs.trans"),as.is=TRUE)
mhtplot2d(d)
```
and ideogram
```{r}
INF1_merge <- read.delim(file.path(INF,"work","INF1.merge"))[c("Chrom","Start","End","prot","MarkerName")]
singletons <- with(INF1_merge,End-Start==1)
INF1_merge[singletons,"Start"] <- INF1_merge[singletons,"Start"] - 1e6
INF1_merge[singletons,"End"] <- INF1_merge[singletons,"End"] + 1e6
small <- with(INF1_merge, Start<0)
INF1_merge[small,"Start"] <- 0
cvt <- read.csv(file.path(INF,"work","/INF1.merge.cis.vs.trans"),as.is=TRUE)
INF1_merge_cvt <- merge(INF1_merge,cvt,by.x=c("prot","MarkerName"),by.y=c("prot","SNP"))
ord <- with(INF1_merge_cvt,order(Chr,bp))
INF1_merge_cvt <- INF1_merge_cvt[ord,]

library(biomaRt)
library(regioneR)
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
bm <- with(INF1_merge_cvt, getBM(attributes = c("hgnc_symbol", "chromosome_name", "start_position", "end_position", "band"),
           filters="chromosomal_region", values=paste0(Chr,":",bp,":",bp), mart=mart))
genes <- with(bm,toGRanges(chromosome_name,start_position,end_position,labels=hgnc_symbol))

library(karyoploteR)
attach(INF1_merge_cvt)
sentinels <- toGRanges(Chr,bp-1,bp,labels=p.gene)
cis.regions <- toGRanges(Chr,cis.start,cis.end)
loci <- toGRanges(Chr,Start,End)
panel <- toGRanges(p.chr,p.start,p.end,labels=p.gene)
colors <- c("red","blue")
seqlevelsStyle(sentinels) <- "UCSC"
kp <- plotKaryotype(genome="hg19",chromosomes=levels(seqnames(sentinels)))
kpAddBaseNumbers(kp)
kpPlotRegions(kp,data=loci,r0=0.05,r1=0.15,border="black")
kpPlotMarkers(kp, data=sentinels, labels=p.gene, text.orientation="vertical",
              cex=0.45, y=0.3*seq_along(p.gene)/length(p.gene), srt=30, ignore.chromosome.ends=TRUE,
              adjust.label.position=TRUE, label.color=colors[2-cis], label.dist=0.002,
              cex.axis=3, cex.lab=3)
legend("bottomright", bty="n", pch=c(19,19), col=colors, pt.cex=0.4, legend=c("cis", "trans"), text.col=colors, cex=0.8, horiz=FALSE)
# kpPlotLinks(kp, data=loci, data2=panel, col=colors[2-cis])
detach(INF1_merge_cvt)
```

Second, we can use pqtlMR for bidirectional Mendelian Randomization via pQTLs.

Third, we can use eQTL Catalogue for colocalization analysis, see example associated with `import_eQTLCatalogue`. A related function is `import_OpenGWAS` used to fetch data from OpenGWAS.

The published work on pQTLs as identified by Sun, et al (2018) and Suhre, et al. (2020) are included as EndNote libraries.

The function `uniprot2ids` converts UniProt IDs to others.

References
==========

Suhre, K., et al. (2020). "Genetics meets proteomics: perspectives for large population-based studies." *Nat Rev Genet*.
	Proteomic analysis of cells, tissues and body fluids has generated valuable insights into the complex processes influencing human biology. Proteins represent intermediate phenotypes for disease and provide insight into how genetic and non-genetic risk factors are mechanistically linked to clinical outcomes. Associations between protein levels and DNA sequence variants that colocalize with risk alleles for common diseases can expose disease-associated pathways, revealing novel drug targets and translational biomarkers. However, genome-wide, population-scale analyses of proteomic data are only now emerging. Here, we review current findings from studies of the plasma proteome and discuss their potential for advancing biomedical translation through the interpretation of genome-wide association analyses. We highlight the challenges faced by currently available technologies and provide perspectives relevant to their future application in large-scale biobank studies.

Sun, B. B., et al. (2018). "Genomic atlas of the human plasma proteome." *Nature* 558(7708): 73-79.
	Although plasma proteins have important roles in biological processes and are the direct targets of many drugs, the genetic factors that control inter-individual variation in plasma protein levels are not well understood. Here we characterize the genetic architecture of the human plasma proteome in healthy blood donors from the INTERVAL study. We identify 1,927 genetic associations with 1,478 proteins, a fourfold increase on existing knowledge, including trans associations for 1,104 proteins. To understand the consequences of perturbations in plasma protein levels, we apply an integrated approach that links genetic variation with biological pathway, disease, and drug databases. We show that protein quantitative trait loci overlap with gene expression quantitative trait loci, as well as with disease-associated loci, and find evidence that protein biomarkers have causal roles in disease using Mendelian randomization analysis. By linking genetic factors to diseases via specific proteins, our analyses highlight potential therapeutic targets, opportunities for matching existing drugs with new disease indications, and potential safety concerns for drugs under development.
